src/main/java/seedu/address/ui/UiManager.java

```diff
@@ -125,7 +128,42 @@ public void setGuiSettings(GuiSettings guiSettings) {
         logic.setGuiSettings(guiSettings);
     };
 
+    public MainWindow getMainWindow() {
+        return this.mainWindow;
+    }
+
+    /**
+     * Passes the string to logic to execute. Gets the command result back from logic, along with any
+     * uiConsumer. If there is one, apply it to the ui
+     * @param commandText String entered by the user
+     * @return  result of the command of class Command result
+     * @throws CommandException
+     * @throws ParseException
+     */
     public CommandResult execute(String commandText) throws CommandException, ParseException {
-        return logic.execute(commandText);
+        CommandResult commandResult = logic.execute(commandText);
+        Consumer<Ui> uiConsumer = commandResult.getUiAction();
+        if (!uiConsumer.equals(null)) {
+            uiConsumer.accept(this);
+        }
```

Comment:
I like your defensive programming! I thought of two things, let me know what you think:

I feel like using the same name (uiAction) consistently across the codebase makes things easier, and in many cases like these, when we use defensive programming against ourselves (in an internal portion of our code), it should probably raise an assertion error rather than being handled silently.

```suggestion
        Consumer<Ui> uiAction = commandResult.getUiAction();
        assert !uiAction.equals(null) : "commandResult.uiAction was set as null";
        if (!uiAction.equals(null)) {
            uiAction.accept(this);
        }
```